### ISSUE TITLE
Unexpected Behavior with `fetchPolicy` "network-only" and `nextFetchPolicy` "cache-first" In 3.7.x

### ISSUE THREAD

**Issue by dairyisscary** [apollographql/apollo-client#10222]
I'm in the middle of doing an upgrade of AC 3.6.8 to 3.7.x (currently tested and template made with 3.7.1) and I found something that I _think_ is a bug, but I'm not sure. 

**Intended outcome:**
I have `useQuery` inside a component that is conditionally shown that uses the `fetchPolicy` of `"network-only"` and `nextFetchPolicy` of `"cache-first"`. In the demo below, I've made the querying component conditionally mount with a toggle button. I expect that every time the component is mounted fresh, it should always fetch from the network and ignore the cache for that first load.

**Actual outcome:**
It appears that, instead, the component comes to life, fetches from the network, fills the cache, then when I hide and fully unmount the component, then when I toggle it back to mounted, it does in fact start fetching from the network but seems to "apply" the `"cache-first"` policy too quickly and displays the cached data _before_ the data comes back from the network. In the demo below, I've raised the delay on the network to 1.5s so it's more obvious and made the length of the people array a count of the number of times the server data is used. 

In other words, I think I expect:
* click to show the querying component, I see `Loading...` for 1.5s, then `1` as the count.
* click to hide the querying component
* click to show the querying component, I see `Loading...` for 1.5s, then `2` as the count.

but instead, what actually happens on that last bullet is I never see `Loading...`, I see `1` as the count, then 1.5s later it updates to `2`.

**How to reproduce the issue:**
https://github.com/dairyisscary/react-apollo-error-template

**Versions**
3.7.x

**Comment by jerelmiller** [comment]
@dairyisscary I played around with this a bit and can indeed confirm the issue you are seeing is a regression. It looks however that this regression was introduced in v3.6.9. I am able to toggle between v3.6.8 to see it work and v3.6.9 to see it break.

Looking at the [changelog](https://github.com/apollographql/apollo-client/blob/main/CHANGELOG.md#apollo-client-369-2022-06-21), it looks like [this PR](https://github.com/apollographql/apollo-client/pull/9823) meant to fix an issue with toggling `skip` also introduced this regression.

**Comment by jerelmiller** [comment]
I've gone ahead and written a [failing test](https://github.com/apollographql/apollo-client/compare/issue-10222-next-fetch-policy?expand=1) that demonstrates this issue.

**Comment by levrik** [comment]
I'm unsure if I'm hit by the same issue but it might be.
Basically I have a custom hook which uses `useQuery` inside. I render the either `data` or `previousData` if `data` is unavailable, so the previous data stays while the new data is loaded. The fetch policy is set to `network-only`.
If now `variables` change to something else and then change back to what they original were, `cache-first` is used, breaking the behavior of rendering the previously fetched data as `data` is not `undefined`.
With Apollo Client 3.5.10 which I used before, `data` was `undefined` and `previousData` was correctly displayed while fetching with the changed variables.
I seem to can work around this by manually setting `nextFetchPolicy` to `network-only` but I'm unsure if this might have other unintended side-effects. No re-mounting involved here.

**Comment by jerelmiller** [comment]
@levrik would you be able to provide a bit more detail on your issue? To make sure I understand, are you saying that you have something like the following?

```js
useQuery(QUERY, {
  fetchPolicy: 'network-only',
  nextFetchPolicy: 'cache-first'
})
```

Some code or a failing test would be tremendously helpful to understand if its a similar issue or not.

One of the changes made in https://github.com/apollographql/apollo-client/pull/9823 is that once a request is kicked off, the `nextFetchPolicy` is then updated synchronously (see more details on [this comment](https://github.com/apollographql/apollo-client/pull/9823#discussion_r897246077) for the reasons why). It's possible you're hitting this case. If not, some code samples would help us understand what you're seeing.

**Comment by levrik** [comment]
@jerelmiller Exactly. I'll try to assemble an example highlighting the issue latest next week.

**Comment by skrivle** [comment]
Hi All 👋 

We are also experiencing this issue it seems. We were able to reproduce this both in a test case and in browser here: https://github.com/skrivle/apollo-next-fetch

This is a regular CRA project. You can run `npm test` and you'll see the test failing. To make the test pass you can comment out the nextFetchPolicy config item in `App.tsx`

For the in browser reproduction please follow these steps:
- `npm start`
- click contact 1
- click contact 2
- click unload (which will unmount the app component)
- click contact 1
- observe that loading is shown very briefly and then it incorrectly shows the cached data for contact 2 until data for contact 1 is loaded.

Note: the conversation items have the same id but different content across the two requests. This is to illustrate the scenario where the properties of an object with a consistent id change over time.

**Comment by JWo1F** [comment]
I have the same bug. When it will be fixed?

**Comment by jerelmiller** [comment]
Hey @JWo1F 👋 !

That is was I [concluded](https://github.com/apollographql/apollo-client/issues/10222#issuecomment-1289378388) as well. The bug was introduced in 3.6.9 (see my links in that comment for more references to the underlying issue). I have a [failing test](https://github.com/apollographql/apollo-client/compare/issue-10222-next-fetch-policy?expand=1#diff-d2c5924a921e0cb39c6d9910863c40eaab64a7901aa9414ab702a6eda79f7e82R4930-R5020) that demonstrates this issue. In fact I attempted to [fix it](https://github.com/apollographql/apollo-client/pull/10239) but so far have been unsuccessful. I'm hoping to come back to this bug sometime soon to get a proper solution in place (if you're curious about more specifics, see my [comment](https://github.com/apollographql/apollo-client/pull/10239#issuecomment-1297726245) on the challenges I've faced thus far).

**Comment by JWo1F** [comment]
Hey @jerelmiller!

Could you tell me, you said you were going to do the task "sometime soon".
When will you have time to fight the bug?

**Comment by JWo1F** [comment]
@phryneas, hi! Do you have any progress on this task?

**Comment by phryneas** [comment]
@JWo1F I am not quite content with the PR yet, and have most of my work directed elsewhere at the moment. I hope that I can return to this within the next few weeks, I'm sorry for the delay!

**Comment by JWo1F** [comment]
@phryneas, hi! Any progress on the task? It's been 3 weeks and we still can't update to the new version... Sorry to keep bugging you, this is a very hot topic for us.

**Comment by phryneas** [comment]
@JWo1F  I'm very sorry that it has been yet another 3 weeks - we have been working hard to get the Next.js support package and the 3.8 beta out, so this has been a bit on the back burner.  

I will be on conference for a week now, but already have a reminder set for this issue once I'm back. I'm really sorry that I can't give you better news than that, but I thought I'd at least keep you in the loop.

Would it help you if I rebase the current PR onto the current `main` branch and make a PR-release that you can use in the meantime?

**Comment by JWo1F** [comment]
@phryneas, yes, it works for us!

**Comment by phryneas** [comment]
@JWo1F I've made a PR release for you - you can install it with `npm i @apollo/client@0.0.0-pr-10631-20230531131910`

**Comment by phryneas** [comment]
@JWo1F I've picked this up again - the PR is in a ready-for-review state now and in a moment, there will be another build available to you over there in a comment :)

**Comment by tatianajiselle** [comment]
I am also seeing a potential memory leak and caching occurring with with fetchPolicy=no-cache in 3.7.9 that im wondering if it could be related to this change? 

In sort we are seeing caching occur despite a no cache policy, when getObsverableByLink() is called in QueryManager.

Not well versed in apollo client code though so hoping someone advise. thx

https://community.apollographql.com/t/client-core-developers-possible-memory-leak-in-apolloclient-querymanager-despite-cache-policy-set-to-no-cache/6390

**Comment by jerelmiller** [comment]
Fixed by #10631 and released with [v3.7.17](https://github.com/apollographql/apollo-client/releases/tag/v3.7.17).
